README.md

```diff
@@ -216,223 +216,59 @@ interface StaticModuleRecord {
-  // tool for a REPL, such that a const or let binding from one evaluation
-  // persists to the next.
-  // Subsequent proposals might afford other modes, like sloppy globals mode.
-  // TODO is sloppyGlobalsMode only sensible in the context of the shim?
-  evaluate(source: string): any;
-
-  // load causes a loader to load module descriptors for the
-  // transitive dependencies of a specified module into its
-  // memo, but does not initialize any modules.
-  // The load function is useful for tools like bundlers and importmap
-  // generators that load a module graph in an emulated host environment
-  // but cannot and should not emulate evaluation.
-  async load(fullSpecifier: string): Promise<void>
-
-  // import induces the asynchronous load phase and then initializes
-  // the given module and any of its transitive dependencies that
-  // have not already begun initialization.
-  async import(fullSpecifier: string): Promise<ModuleExportsNamespace>;
-
-  // TC53: some embedded systems hosts exclude promises and so require these
-  // synchronous variants of import and load.
-  // On hosts where these functions cannot succeed synchronously,
-  // they must throw an error without the side effect of initializing any
-  // additional modules.
-  // For modules that have a top-level await, these would return
-  // after the module first awaits but not after any subsequent promise queue
-  // job.
-
-  // If a host supports both importNow and import, they must share
-  // a memo of full specifiers to promises for the module exports namespace.
-  // importNow and loadNow followed by import or load must not accidentally
-  // reinitialize the underlying module or produce different namespaces.
-
-  importNow(fullSpecifier: string): ModuleExportsNamespace;
-
-  loadNow(fullSpecifier: string): void;
+// ModuleInstance reifies an entangled pair of module environment record
+// and module exports namespace from a particular array of bindings
+// that correspond to the `import` and `export` declarations of a
+// module.
+interface ModuleInstance {
+  // Creates a module instance for either a static module record
+  // or a synthetic static module record.
+  // Creates a module environment record that defers to the
+  // designated global environment record or that associated
+  // with the ModuleInstance constructor.
+  // The module instance will delegate to the given dynamic import
+  // and will use the given global for all eval and Function constructor
+  // behaviors.
+  constructor(
+    record: StaticModuleRecord | SyntheticStaticModuleRecord,
+    {
+      import?: (importSpecifier: string) => Promise<ModuleExportsNamespace>,
+      importMeta?: Object,
+      global?: Global,
+    }
+  );
+
+  // Links the module instance with another module instance
+  // for one of the importSpecifiers declared in the constructed
+  // bindings.
+  link(importSpecifier: string, instance: ModuleInstance);
```

Comment:
This method seems kind've strange given that it expects a specifier but the `ModuleInstance` itself has no way to reflect on what specifiers are requested by the `record` given to the constructor. Or even what the current link status of a given `ModuleInstance` is.

It would be helpful to expose what specifiers still need binding, so that we more easily write a linker:

```ts
async function link(moduleInstance: ModuleInstance) {
    // Need to know what specifiers still need linking so we can
    // recurse and create our loader
    for (const specifier of moduleInstance.unlinkedSpecifiers) {
        const module = await resolveModuleSomehow(specifier, moduleInstance);
        moduleInstance.link(specifier, module);
        // Link child modules
        await link(module);
    }
}
```

Reply:
The ModuleInstance captures the underlying StaticModuleRecord, which in turn has a `bindings` array on which the Loader implementation (either user code or one built into the language), can derive a list of all the necessary linkage and track which one’s it has accounted for. Of course, the ModuleInstance will have to track all of this as well internally and maintain the invariant that a module specifier is only linked once, and that all links are satisfied before it is initialized.

Reply:
> The ModuleInstance captures the underlying StaticModuleRecord, which in turn has a bindings array on which the Loader implementation (either user code or one built into the language), can derive a list of all the necessary linkage and track which one’s it has accounted for. 

It looks like the latest changes make `ModuleInstance` a hidden spec object, so my initial comment is a moot point now.

Reply:
Yes, I think we’re loosely converged on hiding `%ModuleInstance%` at this moment, but I’ve concluded it’s useful to keep and continue to develop the concept as a spec fiction, with the option of revealing it if it receives a lot of support.